### PR TITLE
Add pragma code actions based on GHC warnings

### PR DESCRIPTION
**Pull request by alanz**
Addresses #890.

### CODE REVIEW

File: src/Haskell/Ide/Engine/Plugin/Pragmas.hs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+  where
+    check p =
+      if T.isInfixOf p str
+        then [p]
+        else []
+
+-- ---------------------------------------------------------------------
+
+possiblePragmas :: [T.Text]
+possiblePragmas =
```
Comment (lorenzo) [inline src/Haskell/Ide/Engine/Plugin/Pragmas.hs]: Is this the list of the only extensions GHC suggests? If not, I think we can get the list of extensions from GHC itself.

Comment (alanz) [reply src/Haskell/Ide/Engine/Plugin/Pragmas.hs]: It is the ones that I could find being suggested. 

But I think searching for all of them would be safe, so long as it does not nail performance too much, as it gets checked on every cursor movement.

Part of me says a properly implemented compiled regexp for the match might be a good idea.  Something like an alex lexer with each token being `.*PRAGMATEXT.*`.  But that is probably overkill.

Comment (lorenzo) [reply src/Haskell/Ide/Engine/Plugin/Pragmas.hs]: If this is called constantly for each cursor move, then I think it makes sense to keep the list cached. Either by hardcoding it here of memoizing it after calling the GHC api.

**Comment by lorenzo** [comment]
I think this addresses https://github.com/haskell/haskell-ide-engine/issues/890 and not #843

File: src/Haskell/Ide/Engine/Plugin/Pragmas.hs (hunk 1 above)
Comment (alanz) [reply src/Haskell/Ide/Engine/Plugin/Pragmas.hs]: What is the GHC API call to use? Being lazy ...

Comment (lorenzo) [reply src/Haskell/Ide/Engine/Plugin/Pragmas.hs]: `DynFlags.supportedLanguagesAndExtensions`

**Comment by alanz** [comment]
I wonder if this should not be in the `GhcMod` plugin too, like the unused variable one.

**Comment by lorenzo** [comment]
I kind of like the idea of having more and smaller plugins

**Comment by bubba** [comment]
The GhcMod plugin is getting pretty big. Maybe we should start splitting up some of those code suggestions into plugins too, like `Haskell.Ide.Engine.Plugin.GhcMod.UnusedVariables`,  `Haskell.Ide.Engine.Plugin.GhcMod.Pragmas` etc.

**Comment by alanz** [comment]
Thinking about it some more, having lots of small plugins makes it easier for new contributors to see how it works, and base their contribution on a copy of an existing one.

File: src/Haskell/Ide/Engine/Plugin/Pragmas.hs [hunk 2]
```diff
@@ -0,0 +1,126 @@
+{-# LANGUAGE DeriveAnyClass    #-}
+{-# LANGUAGE DeriveGeneric     #-}
+{-# LANGUAGE OverloadedStrings #-}
+module Haskell.Ide.Engine.Plugin.Pragmas where
```
Comment (bubba) [inline src/Haskell/Ide/Engine/Plugin/Pragmas.hs]: Come to think of this this doesn't really involve GhcMod at all, so this module name should be fine

File: src/Haskell/Ide/Engine/Plugin/Pragmas.hs (hunk 1 above)
Comment (bubba) [reply src/Haskell/Ide/Engine/Plugin/Pragmas.hs]: Output from calling this:

```
["Haskell98","Haskell2010","Unsafe","Trustworthy","Safe","AllowAmbiguousTypes","NoAllowAmbiguousTypes","AlternativeLayoutRule","NoAlternativeLayoutRule","AlternativeLayoutRuleTransitional","NoAlternativeLayoutRuleTransitional","Arrows","NoArrows","AutoDeriveTypeable","NoAutoDeriveTypeable","BangPatterns","NoBangPatterns","BinaryLiterals","NoBinaryLiterals","CApiFFI","NoCApiFFI","CPP","NoCPP","ConstrainedClassMethods","NoConstrainedClassMethods","ConstraintKinds","NoConstraintKinds","DataKinds","NoDataKinds","DatatypeContexts","NoDatatypeContexts","DefaultSignatures","NoDefaultSignatures","DeriveAnyClass","NoDeriveAnyClass","DeriveDataTypeable","NoDeriveDataTypeable","DeriveFoldable","NoDeriveFoldable","DeriveFunctor","NoDeriveFunctor","DeriveGeneric","NoDeriveGeneric","DeriveLift","NoDeriveLift","DeriveTraversable","NoDeriveTraversable","DerivingStrategies","NoDerivingStrategies","DisambiguateRecordFields","NoDisambiguateRecordFields","DoAndIfThenElse","NoDoAndIfThenElse","DoRec","NoDoRec","DuplicateRecordFields","NoDuplicateRecordFields","EmptyCase","NoEmptyCase","EmptyDataDecls","NoEmptyDataDecls","EmptyDataDeriving","NoEmptyDataDeriving","ExistentialQuantification","NoExistentialQuantification","ExplicitForAll","NoExplicitForAll","ExplicitNamespaces","NoExplicitNamespaces","ExtendedDefaultRules","NoExtendedDefaultRules","FlexibleContexts","NoFlexibleContexts","FlexibleInstances","NoFlexibleInstances","ForeignFunctionInterface","NoForeignFunctionInterface","FunctionalDependencies","NoFunctionalDependencies","GADTSyntax","NoGADTSyntax","GADTs","NoGADTs","GHCForeignImportPrim","NoGHCForeignImportPrim","GeneralizedNewtypeDeriving","NoGeneralizedNewtypeDeriving","ImplicitParams","NoImplicitParams","ImplicitPrelude","NoImplicitPrelude","ImpredicativeTypes","NoImpredicativeTypes","IncoherentInstances","NoIncoherentInstances","TypeFamilyDependencies","NoTypeFamilyDependencies","InstanceSigs","NoInstanceSigs","ApplicativeDo","NoApplicativeDo","InterruptibleFFI","NoInterruptibleFFI","JavaScriptFFI","NoJavaScriptFFI","KindSignatures","NoKindSignatures","LambdaCase","NoLambdaCase","LiberalTypeSynonyms","NoLiberalTypeSynonyms","MagicHash","NoMagicHash","MonadComprehensions","NoMonadComprehensions","MonadFailDesugaring","NoMonadFailDesugaring","MonoLocalBinds","NoMonoLocalBinds","MonoPatBinds","NoMonoPatBinds","MonomorphismRestriction","NoMonomorphismRestriction","MultiParamTypeClasses","NoMultiParamTypeClasses","MultiWayIf","NoMultiWayIf","NPlusKPatterns","NoNPlusKPatterns","NamedFieldPuns","NoNamedFieldPuns","NamedWildCards","NoNamedWildCards","NegativeLiterals","NoNegativeLiterals","HexFloatLiterals","NoHexFloatLiterals","NondecreasingIndentation","NoNondecreasingIndentation","NullaryTypeClasses","NoNullaryTypeClasses","NumDecimals","NoNumDecimals","OverlappingInstances","NoOverlappingInstances","OverloadedLabels","NoOverloadedLabels","OverloadedLists","NoOverloadedLists","OverloadedStrings","NoOverloadedStrings","PackageImports","NoPackageImports","ParallelArrays","NoParallelArrays","ParallelListComp","NoParallelListComp","PartialTypeSignatures","NoPartialTypeSignatures","PatternGuards","NoPatternGuards","PatternSignatures","NoPatternSignatures","PatternSynonyms","NoPatternSynonyms","PolyKinds","NoPolyKinds","PolymorphicComponents","NoPolymorphicComponents","PostfixOperators","NoPostfixOperators","QuasiQuotes","NoQuasiQuotes","Rank2Types","NoRank2Types","RankNTypes","NoRankNTypes","RebindableSyntax","NoRebindableSyntax","RecordPuns","NoRecordPuns","RecordWildCards","NoRecordWildCards","RecursiveDo","NoRecursiveDo","RelaxedLayout","NoRelaxedLayout","RelaxedPolyRec","NoRelaxedPolyRec","RoleAnnotations","NoRoleAnnotations","ScopedTypeVariables","NoScopedTypeVariables","StandaloneDeriving","NoStandaloneDeriving","StaticPointers","NoStaticPointers","Strict","NoStrict","StrictData","NoStrictData","TemplateHaskell","NoTemplateHaskell","TemplateHaskellQuotes","NoTemplateHaskellQuotes","TraditionalRecordSyntax","NoTraditionalRecordSyntax","TransformListComp","NoTransformListComp","TupleSections","NoTupleSections","TypeApplications","NoTypeApplications","TypeInType","NoTypeInType","TypeFamilies","NoTypeFamilies","TypeOperators","NoTypeOperators","TypeSynonymInstances","NoTypeSynonymInstances","UnboxedTuples","NoUnboxedTuples","UnboxedSums","NoUnboxedSums","UndecidableInstances","NoUndecidableInstances","UndecidableSuperClasses","NoUndecidableSuperClasses","UnicodeSyntax","NoUnicodeSyntax","UnliftedFFITypes","NoUnliftedFFITypes","ViewPatterns","NoViewPatterns"]
```

Are only of these pragma-ble?

**Comment by alanz** [comment]
I have no idea.

I think it might be better to keep the list small, and add to it via complaints, otherwise it will kill performance.

**Pull request merged**